lib/rust/profiling/src/lib.rs

```diff
@@ -0,0 +1,512 @@
+    start_interval(metadata.clone()).release();
+    let value = closure();
+    let measurement = end_interval(metadata);
+
+    IntervalMeasurementResult { value, measurement }
+}
+
+/// Result of profiling a closure via `measure_interval`. Contains the measurement result and the
+/// closure return value.
+#[derive(Clone, Debug)]
+pub struct IntervalMeasurementResult<T> {
+    /// Return value of the measured closure.
+    pub value:       T,
+    /// Measurement result.
+    pub measurement: Result<Measurement, MeasurementError>,
+}
+
+
+
+// ======================
+// === IntervalHandle ===
+// ======================
+
+/// Handle that allows ending the interval.
+#[derive(Clone, Debug)]
+pub struct IntervalHandle {
+    metadata: Metadata,
+    released: bool,
+}
+
+impl IntervalHandle {
+    fn new(metadata: Metadata) -> Self {
+        IntervalHandle { metadata, released: false }
+    }
+
+    /// Measure the interval.
+    pub fn end(mut self) {
+        self.released = true;
+        warn_on_error(end_interval(self.metadata.clone()));
+    }
+
+    /// Release the handle to prevent a warning to be emitted when it is garbage collected without
+    /// a call to `end`. This can be useful if one wants to call `end_interval` manually, or the
+    /// equivalent call to `end_interval` is in Rust code.
+    pub fn release(mut self) {
+        self.released = true;
+        drop(self)
+    }
+}
+
+impl Drop for IntervalHandle {
+    fn drop(&mut self) {
+        if !self.released {
+            warn_on_error(end_interval(self.metadata.clone()));
+            WARNING!(format!("{} was dropped without a call to `measure`.", self.metadata.label));
+        }
+    }
+}
+
+
+
+// ===============
+// === Reports ===
+// ===============
+
+/// Report of all measurements taken during the current session.
+#[derive(Clone, Debug)]
+pub struct Report {
+    measures: Vec<Measurement>,
+}
+
+impl Report {
+    /// Create a new report from the measurements registered in the Performance API.
+    pub fn generate() -> Self {
+        let js_measures = performance().get_entries_by_type("measure");
+
+        let mut measures = Vec::default();
+
+        js_measures.for_each(&mut |entry, _, _| {
+            let entry: PerformanceEntry = entry.into();
+            let measurement: Result<Measurement, _> = entry.try_into();
+            if let Ok(measurement) = measurement {
+                measures.push(measurement)
+            }
+        });
+        Report { measures }
+    }
+}
+
+impl Display for Report {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        let sorted_measurements =
+            self.measures.iter().sorted_by_key(|measure| OrderedFloat(measure.start_time));
+        for to_print in sorted_measurements {
+            f.write_str(&format!("{}\n", &to_print))?;
+        }
+        Ok(())
+    }
+}
+
+/// Return a report on all recorded measurement entries.
+pub fn entries() -> Report {
+    Report::generate()
+}
```

Comment:
No tests – add tests showing the usage of the library, please. They do not have to be extensive, just show how to use different modes there, please.